### DISCUSSION TITLE
solver.step can not update the parameters

### DISCUSSION THREAD

**Discussion by fireyan8** [pybamm-team/PyBaMM#3416] [Q&A] [answered]
Hallo everyone,
 I am using solver.step to update current for each step. But it seems not working.
here is the code
```


model = pybamm.lithium_ion.DFN(options)
params = pybamm.ParameterValues("Ecker2015").copy()
params.update({"Current function [A]": 0.0,'})
initial_soc = 0.2

params.set_initial_stoichiometries(initial_soc)
model1 = params.process_model(model, inplace=False)

geometry = model.default_geometry
submesh_types = model.default_submesh_types
var_pts = model.default_var_pts
params.process_geometry(geometry)
mesh = pybamm.Mesh(geometry, submesh_types, var_pts)

spatial_methods = model.default_spatial_methods
disc = pybamm.Discretisation(mesh, spatial_methods)
disc_model = disc.process_model(model=model1, inplace=False)




solver = pybamm.CasadiSolver(mode="safe", atol=1e-6, rtol=1e-3)


initial_solution = None


dt = 1e-1
total_time = 1  


inputs = {name: -1 for name in model.variable_names()}
print(inputs)

current_time = 0
while current_time < total_time:
    solution = solver.step(old_solution=initial_solution, model=disc_model, dt=dt, inputs=inputs)
    initial_solution = solution
    current_time += dt

initial_solution.plot()
```
after ploting the current is always 0 at every step.

what should be given to the inputs for solver.step()

I am using
python 3.11
pybamm 23.5
windows 11

thank you for your help.

update: 
I found this in Pybamm test_casadi_solver.py:
```
    def test_model_step_with_input(self):
        # Create model
        model = pybamm.BaseModel()
        var = pybamm.Variable("var")
        a = pybamm.InputParameter("a")
        model.rhs = {var: a * var}
        model.initial_conditions = {var: 1}
        model.variables = {"a": a}

        solver = pybamm.CasadiSolver(rtol=1e-8, atol=1e-8)

        # Step with an input
        dt = 0.1
        step_sol = solver.step(None, model, dt, npts=5, inputs={"a": 0.1})
        np.testing.assert_array_equal(step_sol.t, np.linspace(0, dt, 5))
        np.testing.assert_allclose(step_sol.y.full()[0], np.exp(0.1 * step_sol.t))

        # Step again with different inputs
        step_sol_2 = solver.step(step_sol, model, dt, npts=5, inputs={"a": -1})
        np.testing.assert_array_almost_equal(
            step_sol_2.t,
            np.array([0, 0.025, 0.05, 0.075, 0.1, 0.1 + 1e-9, 0.125, 0.15, 0.175, 0.2]),
        )
        np.testing.assert_array_equal(
            step_sol_2["a"].entries,
            np.array([0.1, 0.1, 0.1, 0.1, 0.1, -1, -1, -1, -1, -1]),
        )
        np.testing.assert_allclose(
            step_sol_2.y.full()[0],
            np.concatenate(
                [
                    np.exp(0.1 * step_sol_2.t[:5]),
                    np.exp(0.1 * step_sol_2.t[4])
                    * np.exp(-(step_sol_2.t[5:] - step_sol_2.t[5])),
                ]
            ),
        )
```

which shows the solver.step() is working on casadisolver.   How to let it also work on DFN or SPM model?

**Comment by TomTranter** [comment]
It doesn't look you are changing the inputs inside the loop

**Reply by fireyan8** [reply]
Hi,
The original current is 0, the input was set to -1. When I plot. It still shows 0.  Which means the Inputs are not working

**Reply by TomTranter** [reply]
Follow this example for proper use of inputs https://docs.pybamm.org/en/latest/source/examples/notebooks/parameterization/change-input-current.html

**Reply by fireyan8** [reply]
In this example, it is not solving the model stepwise. I want to solve the model step by step and changing the current for each step.
Before I did it by setting up a new casadisolver for each step. But solver.step() should also do the same without copying all the matrixs,
however the inputs seems not working.

**Reply by TomTranter** [reply]
You haven't defined the input correctly

**Reply by fireyan8** [reply]
HI:      
from the base_solver.py
step():
  inputs : dict, optional
            Any input parameters to pass to the model when solving

so my inputs  should:

```
inputs = {"Current function": -1}
```
is that right?

**Reply by TomTranter** [reply]
Yes but the definition of the current in parameter values isn't

**Reply by fireyan8** [reply]
So what is the correct to use solver.step()?

**Reply by TomTranter** [reply]
Solver step has nothing to do with it. Read the example I posted

**Reply by fireyan8** [reply]
From the example,
if i want to do step by step.
model = pybamm.lithium_ion.DFN()
solutions = []
# set the default model parameters
param = model.default_parameter_values

# change the current function to be an input parameter
param["Current function [A]"] = "[input]"
# set up simlation
simulation = pybamm.Simulation(model, parameter_values=param)

# solve the model at the given time points, passing the current as an input
t_eval1 = np.linspace(0, 1, 3)
t_eval2 = np.linspace(1, 2, 3)
s1 = simulation.solve(t_eval1, inputs={"Current function [A]": 1.6})
s2 = simulation.solve(t_eval2, inputs={"Current function [A]": -1.6})
solutions.append(s1)
solutions.append(s2)
# plot
s1['Current [A]'].data

the results are not continuous.

please tell me the correct code to do it stepwise.

**Reply by TomTranter** [reply, marked as answer]
Something like this should work
```
import pybamm
model = pybamm.lithium_ion.DFN()

# set the default model parameters
param = model.default_parameter_values

# change the current function to be an input parameter
param["Current function [A]"] = "[input]"

# set up simlation
simulation = pybamm.Simulation(model, parameter_values=param)

# solve the model at the given time points, passing the current as an input
simulation.step(dt=100, npts=100, inputs={"Current function [A]": 1.6})
simulation.step(dt=100, npts=100, inputs={"Current function [A]": -1.6})

# plot
simulation.plot()
```

**Reply by fireyan8** [reply]
thank you very much
